In [3]:
import pandas as pd
import numpy as np

## Import the data

In [2]:
excel_data = pd.ExcelFile('data_task.xlsx')

In [4]:
#check the sheets names
excel_data.sheet_names

['Instructions', 'FY18 - loans', 'FY19 - loans', 'FY19 - collaterals']

In [5]:
sheet1 = excel_data.parse('FY18 - loans', skiprows = 2, columns=[0,1])
sheet2 = excel_data.parse('FY19 - loans', skiprows = 2, columns=[0,1])
sheet3 = excel_data.parse('FY19 - collaterals', skiprows = 2, columns=[0,1])

In [7]:
sheet1, sheet2, sheet3

(     Loan ID  Outstanding balance (kEUR)
 0     100001                      1795.0
 1     100002                      1954.0
 2     100003                      1965.0
 3     100004                      1217.0
 4     100005                      1425.0
 ...      ...                         ...
 996   100997                      1630.0
 997   100998                      1108.0
 998   100999                      1010.0
 999   101000                      1630.0
 1000   TOTAL                   1510888.0
 
 [1001 rows x 2 columns],
      Loan ID Outstanding balance (mEUR)
 0     100001                      1.755
 1     100002                      1.928
 2     100003                      1.945
 3     100004                      1.122
 4     100005                      1.396
 ...      ...                        ...
 996   101007                      1.968
 997   101008                      1.539
 998   101009                      1.145
 999   101010                      1.466
 1000   TOTAL    

## Merge tables in sheet1 & sheet2 based on Loan ID

In [8]:
sheet_1_2 = sheet1.merge(sheet2, on='Loan ID', how='outer', sort=True)
sheet_1_2

Loan ID  Outstanding balance (kEUR) Outstanding balance (mEUR)
0     100001                      1795.0                      1.755
1     100002                      1954.0                      1.928
2     100003                      1965.0                      1.945
3     100004                      1217.0                      1.122
4     100005                      1425.0                      1.396
...      ...                         ...                        ...
1007  101007                         NaN                      1.968
1008  101008                         NaN                      1.539
1009  101009                         NaN                      1.145
1010  101010                         NaN                      1.466
1011   TOTAL                   1510888.0                   1404.673

[1012 rows x 3 columns]

In [9]:
# General check of missing data
sheet_1_2.isna().sum() 

Loan ID                        0
Outstanding balance (kEUR)    11
Outstanding balance (mEUR)    11
dtype: int64

## How many loans were in the portfolio in FY18 which were missing in FY19? What are their IDs

In [12]:
nan_FY19 = sheet_1_2[sheet_1_2['Outstanding balance (mEUR)'].isnull()] #missing data in FY19
nan_FY19, nan_FY19.isna().sum()

(    Loan ID  Outstanding balance (kEUR) Outstanding balance (mEUR)
 195  100196                      1346.0                        NaN
 290  100291                      1823.0                        NaN
 307  100307                      1490.0                        NaN
 350  100350                      1950.0                        NaN
 351  100351                      1968.0                        NaN
 468  100468                      1666.0                        NaN
 584  100584                      1889.0                        NaN
 635  100635                      1380.0                        NaN
 704  100704                      1413.0                        NaN
 896  100896                      1902.0                        NaN
 951  100951                      1218.0                        NaN,
 Loan ID                        0
 Outstanding balance (kEUR)     0
 Outstanding balance (mEUR)    11
 dtype: int64)

In [16]:
np.array(nan_FY19['Loan ID'])

array([100196, 100291, 100307, 100350, 100351, 100468, 100584, 100635,
       100704, 100896, 100951], dtype=object)

###### <span style='color:blue'>11 loans were in the portfolio in FY18 which were missing in FY19<span>
    
###### <span style='color:blue'>Their IDs: 100196, 100291, 100307, 100350, 100351, 100468, 100584, 100635,100704, 100896, 100951

# -------------------------------------------------------------------------------------------------------------

## How many loans were in the portfolio in FY19 which were missing in FY18? What are their IDs

In [17]:
nan_FY18 = sheet_1_2[sheet_1_2['Outstanding balance (kEUR)'].isnull()] #missing data in FY19
nan_FY18, nan_FY18.isna().sum()

(     Loan ID  Outstanding balance (kEUR) Outstanding balance (mEUR)
 681   100681                         NaN                      1.285
 1001  101001                         NaN                      1.756
 1002  101002                         NaN                      1.077
 1003  101003                         NaN                      1.133
 1004  101004                         NaN                      1.364
 1005  101005                         NaN                      1.645
 1006  101006                         NaN                       1.61
 1007  101007                         NaN                      1.968
 1008  101008                         NaN                      1.539
 1009  101009                         NaN                      1.145
 1010  101010                         NaN                      1.466,
 Loan ID                        0
 Outstanding balance (kEUR)    11
 Outstanding balance (mEUR)     0
 dtype: int64)

In [18]:
np.array(nan_FY18['Loan ID'])

array([100681, 101001, 101002, 101003, 101004, 101005, 101006, 101007,
       101008, 101009, 101010], dtype=object)

###### <span style='color:blue'>11 loans were in the portfolio in FY19 which were missing in FY18<span>
###### <span style='color:blue'>Their IDs: 100681, 101001, 101002, 101003, 101004, 101005, 101006, 101007, 101008, 101009, 101010

# -------------------------------------------------------------------------------------------------------

## How much did the portfolio grow or decline between FY18 and FY19 in outstanding amounts?

In [21]:
#Growth_amount = Total Outstanding balance in 2019 - Total Outstanding balance in 2018
#Growth_ratio % = (Growth_amount/Total Outstanding balance in 2018)*100

growth_amount = sheet_1_2['Outstanding balance (mEUR)'][1011] * 1000 - sheet_1_2['Outstanding balance (kEUR)'][1011]
growth_ratio = growth_amount/sheet_1_2['Outstanding balance (kEUR)'][1011]*100
print(growth_amount,'kEUR'), print(growth_ratio, '%')


-106214.99999999977 kEUR
-7.029971778186058 %


(None, None)

###### <span style='color:blue'>The portffoilio decline was 106214 kEUR, with a percentage ratio around 7%

# -------------------------------------------------------------------------------------------------------------

## What is the highest loan-to-value figure in the portfolio in FY19 (calculated via outstanding balance of the loan divided by the sum of the collateral asset values related to the loan)?

In [22]:
# check duplicated IDs 
sheet3['Loan ID'].duplicated().sum()

168

In [25]:
sheet2['Loan ID'].duplicated().sum()

1

In [27]:
# check which loan is duplicated
sheet2[sheet2['Loan ID'].duplicated()]

Loan ID Outstanding balance (mEUR)
304  100306                     -0.863

In [28]:
np.where(sheet2['Loan ID'] == 100306)

(array([303, 304], dtype=int64),)

In [30]:
sheet2['Outstanding balance (mEUR)'][303], sheet2['Outstanding balance (mEUR)'][304]

(0.863, -0.863)

In [31]:
# Loan ID 100306 is duplicated twice in FY19. the two values of (Outstanding balance (mEUR)) neglect eachother in case of aggregating them.
#I choose to don't consider this ID as one velue are suspecious, especially the negative value.
sheet_2 = sheet2.drop(index=[303,304])

In [32]:
# sum of the collateral asset values related to the loan
agg = pd.DataFrame(sheet3.groupby(['Loan ID'])['Asset value (mEUR)'].sum())
agg

Asset value (mEUR)
Loan ID                    
100001                2.903
100002                0.855
100003                2.850
100004                1.997
100005                1.015
...                     ...
101007                2.808
101008                4.218
101009                2.953
101010                2.935
TOTAL              1812.801

[885 rows x 1 columns]

In [34]:
1053-885 

168

In [37]:
# merge the outstanding balance of the loan with the sum of the collateral asset values 
new_sheet3 = agg.merge(sheet_2, on='Loan ID', how='outer', sort=True)

In [38]:
new_sheet3

Loan ID  Asset value (mEUR) Outstanding balance (mEUR)
0    100001               2.903                      1.755
1    100002               0.855                      1.928
2    100003               2.850                      1.945
3    100004               1.997                      1.122
4    100005               1.015                      1.396
..      ...                 ...                        ...
994  101007               2.808                      1.968
995  101008               4.218                      1.539
996  101009               2.953                      1.145
997  101010               2.935                      1.466
998   TOTAL            1812.801                   1404.673

[999 rows x 3 columns]

In [39]:
new_sheet3.isnull().sum()

Loan ID                         0
Asset value (mEUR)            114
Outstanding balance (mEUR)      0
dtype: int64

In [49]:
new_sheet3[0:998]

Loan ID  Asset value (mEUR) Outstanding balance (mEUR)
0    100001               2.903                      1.755
1    100002               0.855                      1.928
2    100003               2.850                      1.945
3    100004               1.997                      1.122
4    100005               1.015                      1.396
..      ...                 ...                        ...
993  101006               1.349                       1.61
994  101007               2.808                      1.968
995  101008               4.218                      1.539
996  101009               2.953                      1.145
997  101010               2.935                      1.466

[998 rows x 3 columns]

In [50]:
new_sheet3['LTV'] = new_sheet3['Outstanding balance (mEUR)'][0:998]/new_sheet3['Asset value (mEUR)'][0:998] 

In [52]:
new_sheet3['LTV'].max()

3.753846153846154

###### <span style='color:blue'>The highest loan-to-value (LTV) in the portfolio in FY19 is: 3.754, with a ratio:375.4%

# --------------------------------------------------------------------------------------------------------

## Is there a notice of anthing unusual in the data

In [78]:
#Calculate some summary statistics of (Outstanding balance (mEUR)) in sheet1 (FY18)
mean1 = sheet1['Outstanding balance (kEUR)'][0:1000].mean()
stdv1 = sheet1['Outstanding balance (kEUR)'][0:1000].std()
#Z_score
sheet1['Z'] = (sheet1['Outstanding balance (kEUR)'][0:1000]-mean1)/stdv1

In [80]:
sheet1[0:1000]

Loan ID  Outstanding balance (kEUR)         Z
0    100001                      1795.0  0.995021
1    100002                      1954.0  1.551961
2    100003                      1965.0  1.590492
3    100004                      1217.0 -1.029579
4    100005                      1425.0 -0.301003
..      ...                         ...       ...
995  100996                      1740.0  0.802369
996  100997                      1630.0  0.417064
997  100998                      1108.0 -1.411381
998  100999                      1010.0 -1.754652
999  101000                      1630.0  0.417064

[1000 rows x 3 columns]

In [58]:
#Calculate some summary statistics of (Outstanding balance (mEUR)) in sheet1 (FY19)

#Sheet_2(Outstanding balance (mEUR)) contains strings
#Check its index
for i in range(0,999):
    if type(sheet2['Outstanding balance (mEUR)'][i]) == str:
        print(i)

858


In [59]:
sheet2['Outstanding balance (mEUR)'][858]

'1.148m'

In [60]:
#drop the row that contains string 
new_sheet_2 = sheet_2.drop(index=[858])

In [82]:
#Calculate some summary statistics of (Outstanding balance (mEUR)) in sheet1 (FY19)
mean2 = new_sheet_2['Outstanding balance (mEUR)'][0:997].mean()
stdv2 = new_sheet_2['Outstanding balance (mEUR)'][0:997].std()
#Z_score
new_sheet_2['Z'] = (new_sheet_2['Outstanding balance (mEUR)'][0:997]-mean2)/stdv2

In [84]:
new_sheet_2[0:997]

Loan ID Outstanding balance (mEUR)         Z
0    100001                      1.755  1.196033
1    100002                      1.928  1.791893
2    100003                      1.945  1.850446
3    100004                      1.122 -0.984197
4    100005                      1.396 -0.040464
..      ...                        ...       ...
995  101006                       1.61  0.696612
996  101007                      1.968  1.929665
997  101008                      1.539  0.452068
998  101009                      1.145 -0.904979
999  101010                      1.466  0.200635

[997 rows x 3 columns]

In [85]:
#Calculate some summary statistics of (Outstanding balance (mEUR)) in sheet1 (FY19)
mean3 = sheet3['Asset value (mEUR)'][0:1052].mean()
stdv3 = sheet3['Asset value (mEUR)'][0:1052].std()
#Z_score
sheet3['Z'] = (sheet3['Asset value (mEUR)']-mean3)[0:1052]/stdv3

In [87]:
sheet3[0:1052]

Loan ID  Asset ID  Asset value (mEUR)         Z
0     100001       1.0               2.903  1.595344
1     100002       2.0               0.855 -1.210773
2     100003       3.0               2.850  1.522725
3     100004       4.0               1.997  0.353966
4     100005       5.0               1.015 -0.991545
...      ...       ...                 ...       ...
1047  100260    1048.0               1.791  0.071710
1048  100310    1049.0               2.033  0.403292
1049  100311    1050.0               2.022  0.388220
1050  100315    1051.0               1.367 -0.509244
1051  100316    1052.0               1.890  0.207357

[1052 rows x 4 columns]

In [ ]:
#The ideal range of Z score is between -2 & 2

In [97]:
print('min1:', sheet1['Z'][0:1000].min(),'max1:',sheet1['Z'][0:1000].max()),
print('min2:',new_sheet_2['Z'][0:997].min(),'max2:',new_sheet_2['Z'][0:997].max()),
print('min3:',sheet3['Z'][0:1052].min(),'max3:',sheet3['Z'][0:1052].max())

min1: -1.786176744904511 max1: 1.7130884582595525
min2: -1.993371772082659 max2: 1.929664658800825
min3: -1.695814794046452 max3: 1.7255105882068171


###### <span style='color:blue'>Since the Z-score for each data-set is confined between the values -2 and 2, I don't notice anything unusual in the data